br/pkg/restore/ingestrec/ingest_recorder.go

```diff
@@ -0,0 +1,182 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingestrec
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/types"
+)
+
+// IngestIndexInfo records the information used to generate index drop/re-add SQL.
+type IngestIndexInfo struct {
+	SchemaName string
+	TableName  string
+	ColumnList string
+	IsPrimary  bool
+	IndexInfo  *model.IndexInfo
+	Updated    bool
+}
+
+// IngestRecorder records the indexes information that use ingest mode to construct kvs.
+// Currently log backup cannot backed up these ingest kvs. So need to re-construct them.
+type IngestRecorder struct {
+	// Table ID -> Index ID -> Index info
+	items map[int64]map[int64]*IngestIndexInfo
+}
+
+// Return an empty IngestRecorder
+func New() *IngestRecorder {
+	return &IngestRecorder{
+		items: make(map[int64]map[int64]*IngestIndexInfo),
+	}
+}
+
+func notIngestJob(job *model.Job) bool {
+	return job.ReorgMeta == nil ||
+		job.ReorgMeta.ReorgTp != model.ReorgTypeLitMerge
+}
+
+func notAddIndexJob(job *model.Job) bool {
+	/* support new index using accelerated indexing in future:
+	 * // 1. skip if the new index didn't generate new kvs
+	 * // 2. shall the ReorgTp of ModifyColumnJob be ReorgTypeLitMerge if use accelerated indexing?
+	 * if job.RowCount > 0 && notIngestJob(job) {
+	 *   // ASSERT: select new indexes, which have the highest IDs in this job's BinlogInfo
+	 *   newIndexesInfo := getIndexesWithTheHighestIDs(len(indexIDs))
+	 *   for _, newIndexInfo := range newIndexesInfo {
+	 *     tableindexes[newIndexInfo.ID] = ...
+	 *   }
+	 * }
+	 */
+	return job.Type != model.ActionAddIndex &&
+		job.Type != model.ActionAddPrimaryKey
+}
+
+func notSynced(job *model.Job) bool {
+	return job.State != model.JobStateSynced
+}
+
+// AddJob firstly filters the ingest index add operation job, and records it into IngestRecorder.
+func (i *IngestRecorder) AddJob(job *model.Job) error {
+	if job == nil || notIngestJob(job) || notAddIndexJob(job) || notSynced(job) {
+		return nil
+	}
+
+	// If the add-index operation affects no row, the index doesn't need to be repair.
+	if job.RowCount == 0 {
+		return nil
+	}
+
+	var indexID int64 = 0
+	if err := job.DecodeArgs(&indexID); err != nil {
+		return errors.Trace(err)
+	}
+
+	tableindexes, exists := i.items[job.TableID]
+	if !exists {
+		tableindexes = make(map[int64]*IngestIndexInfo)
+		i.items[job.TableID] = tableindexes
+	}
+
+	// the current information of table/index might be modified by other ddl jobs,
+	// therefore update the index information at last
+	tableindexes[indexID] = &IngestIndexInfo{
+		IsPrimary: job.Type == model.ActionAddPrimaryKey,
+		Updated:   false,
+	}
```

Comment:
Why not record `job.Query` as the SQL?

Reply:
Beacuse the record `job.Query` is not the final information.
In the upstream cluster, we first create an index by `ALTER TABLE test.pairs ADD INDEX i1(x, y)`.
Then we rename something such like `RENAME TABLE test.pairs TO test.pairs2`.

When we do restore, the ingest job's `Query` we get is `ALTER TABLE test.pairs ADD INDEX i1(x, y)`. However, the table `test.pairs` is not exists currently.